pennylane/decomposition/decomposition_graph.py

```diff
@@ -0,0 +1,243 @@
+
+        Args:
+            op (Operator): The operator for which to return the resource estimates.
+
+        Returns:
+            Resources: The resource estimates.
+
+        """
+        op_node = resource_rep(type(op), **op.resource_params)
+        op_node_idx = self._op_node_indices[op_node]
+        return self._visitor.d[op_node_idx]
+
+    def decomposition(self, op) -> DecompositionRule:
+        """Returns the optimal decomposition rule for a given operator.
+
+        Args:
+            op (Operator): The operator for which to return the optimal decomposition.
+
+        Returns:
+            DecompositionRule: The optimal decomposition.
+
+        """
+        op_node = resource_rep(type(op), **op.resource_params)
+        op_node_idx = self._op_node_indices[op_node]
+        d_node_idx = self._visitor.p[op_node_idx]
+        return self._graph[d_node_idx].rule
+
+
+class _DecompositionSearchVisitor(DijkstraVisitor):
+    """The visitor used in the dijkstra search for the optimal decomposition."""
+
+    def __init__(self, graph: rx.PyDiGraph, original_op_indices: set[int], lazy: bool = True):
+        self._graph = graph
+        self._lazy = lazy
+        self.d: dict[int, Resources] = {}  # maps node indices to the optimal resource estimates
+        self.p: dict[int, int] = {}  # maps operator nodes to the optimal decomposition nodes
+        self.unsolved_op_indices = original_op_indices.copy()
+        self._num_edges_examined: dict[int, int] = {}  # keys are decomposition node indices
+
+    def edge_weight(self, edge_obj):
+        """Calculates the weight of an edge."""
+        if not isinstance(edge_obj, tuple):
+            return float(edge_obj)
+        op_node_idx, d_node_idx = edge_obj
+        return self.d[d_node_idx].num_gates - self.d[op_node_idx].num_gates
+
+    def discover_vertex(self, v, _):
+        """Triggered when a vertex is about to be explored during the dijkstra search."""
+        self.unsolved_op_indices.discard(v)
+        if not self.unsolved_op_indices and self._lazy:
+            raise StopSearch
+
+    def examine_edge(self, edge):
+        """Triggered when an edge is examined during the dijkstra search."""
+        src_idx, target_idx, _ = edge
+        src_node = self._graph[src_idx]
+        target_node = self._graph[target_idx]
+        if not isinstance(target_node, _DecompositionNode):
+            return  # nothing is to be done for edges leading to an operator node
+        if target_idx not in self.d:
+            self.d[target_idx] = Resources()  # initialize with empty resource
+        self.d[target_idx] += self.d[src_idx] * target_node.count(src_node)
+        if target_idx not in self._num_edges_examined:
+            self._num_edges_examined[target_idx] = 0
+        self._num_edges_examined[target_idx] += 1
+        if self._num_edges_examined[target_idx] < len(target_node.resource_decomp.gate_counts):
+            raise PruneSearch
```

Comment:
I might have gotten lost here, but what exactly is the logic here? Might be nice to have a dev comment :)

Reply:
Done.